File: src-python/leaderboard.py
class Leaderboard:
    def __init__(self) -> None:
        self.scores: dict[str, int] = {}

    def add_score(self, username: str, points: int):
        if username in self.scores:
            self.scores[username] += points
        else:
            self.scores[username] = points

    def remove_score(self, username: str, points: int):
        if username in self.scores:
            self.scores[username] -= points
            if self.scores[username] <= 0:
                del self.scores[username]

    def get_score(self, username: str) -> int:
        return self.scores.get(username, 0)

    def display_leaderboard(self):
        print("Leaderboard:")
        sorted_scores = sorted(self.scores.items(), key=lambda x: x[1], reverse=True)
        for i, (username, points) in enumerate(sorted_scores, start=1):
            print(f"{i}. {username}: {points}")

File: src/lib/game/leaderboard.ts
export class Leaderboard {
	scores: Map<string, number>;

	constructor() {
		this.scores = new Map();
	}

	reset(): void {
		this.scores = new Map();
	}

	addScore(username: string, points: number): void {
		const score = this.scores.get(username);
		this.scores.set(username, (score || 0) + points);
	}

	removeScore(username: string): boolean {
		return this.scores.delete(username);
	}

	getScore(username: string): number {
		return this.scores.get(username) || 0;
	}

	getLeaderboard() {
		const leaderboard = [];
		for (const [index, key] of Array.from(this.scores)
			.sort((a, b) => b[1] - a[1])
			.entries()) {
			leaderboard.push({ username: key[0], score: key[1] });
			if (index > 10) {
				break;
			}
		}
		return leaderboard;
	}
}
